File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/routing/FocusNavigationPolicy.kt
package org.bilibilifocus.core.routing

import org.bilibilifocus.core.model.FocusSettings

class FocusNavigationPolicy(settings: FocusSettings = FocusSettings.DEFAULTS) {

    sealed interface Decision {
        data object Allow : Decision
        data object Cancel : Decision
        data class Redirect(val url: String) : Decision
    }

    fun decision(url: String, currentURL: String? = null): Decision {
        val canonicalURL = canonicalWebURL(url)
        if (canonicalURL != url) {
            return Decision.Redirect(canonicalURL)
        }

        if (shouldBlock(canonicalURL, currentURL)) {
            return Decision.Cancel
        }

        return Decision.Allow
    }

    private fun shouldBlock(url: String, currentURL: String?): Boolean {
        val parsed = BiliUrl.parse(url)
        val scheme = parsed.scheme?.lowercase() ?: ""
        if (scheme in blockedSchemes) return true

        val host = parsed.host ?: ""
        if (host in blockedHosts) return true

        val lowercasedPath = parsed.path.lowercase()
        val lowercasedURL = url.lowercase()
        val isBilibiliFamilyHost = "bilibili.com" in host || "hdslb.com" in host
        val cameFromVideoPage = currentURL?.let {
            val currentParsed = BiliUrl.parse(it)
            currentParsed.path.startsWith("/video/") ||
                (currentParsed.host?.contains("bilibili.com") == true &&
                    it.lowercase().contains("/video/"))
        } ?: false

        if (isBilibiliFamilyHost && lowercasedPath.contains("/download")) return true
        if (lowercasedURL.contains("openapp") || lowercasedURL.contains("launchapp")) return true
        if (cameFromVideoPage && isBilibiliFamilyHost && lowercasedURL.contains("download")) return true

        return false
    }

    companion object {
        private val blockedSchemes = setOf(
            "bilibili", "bilibilihd", "bstar", "intent", "itms-apps", "itmss",
        )
        private val blockedHosts = setOf("app.bilibili.com")

        fun canonicalWebURL(url: String): String {
            val parsed = BiliUrl.parse(url)
            val host = parsed.host ?: return url

            val path = parsed.path.lowercase()
            val isStandardBilibiliHost =
                host == "www.bilibili.com" || host == "bilibili.com" || host == "m.bilibili.com"

            // Normalize video/bangumi hosts to www.bilibili.com
            if (isStandardBilibiliHost &&
                (path.startsWith("/video/") || path.startsWith("/bangumi/play/"))
            ) {
                return parsed.withHost("www.bilibili.com").toUrl()
            }

            // Convert blackboard player URLs to canonical video URLs
            if (isStandardBilibiliHost &&
                (path.startsWith("/blackboard/html5player.html") ||
                    path.startsWith("/blackboard/html5mobileplayer.html"))
            ) {
                val videoPath = canonicalVideoPath(parsed)
                if (videoPath != null) {
                    val filteredItems = filteredVideoQueryItems(parsed.queryItems)
                    return parsed
                        .withScheme("https")
                        .withHost("www.bilibili.com")
                        .withPath(videoPath)
                        .withQueryItems(filteredItems)
                        .withFragment(null)
                        .toUrl()
                }
            }

            // Normalize opus paths
            if (isStandardBilibiliHost && path.startsWith("/opus/")) {
                return parsed.withHost("www.bilibili.com").toUrl()
            }

            // Convert t.bilibili.com/NNNN to www.bilibili.com/opus/NNNN
            if (host == "t.bilibili.com") {
                val pathComponents = parsed.path.split("/").filter { it.isNotEmpty() }
                if (pathComponents.size == 1 && pathComponents[0].all { it.isDigit() }) {
                    return parsed
                        .withScheme("https")
                        .withHost("www.bilibili.com")
                        .withPath("/opus/${pathComponents[0]}")
                        .toUrl()
                }
            }

            // Normalize live room URLs
            if (host == "live.bilibili.com") {
                val roomID = canonicalLiveRoomID(parsed)
                if (roomID != null) {
                    return parsed
                        .withScheme("https")
                        .withHost("live.bilibili.com")
                        .withPath("/$roomID")
                        .withQueryItems(null)
                        .withFragment(null)
                        .toUrl()
                }
            }

            return url
        }

        private fun canonicalVideoPath(parsed: BiliUrl): String? {
            val bvid = parsed.queryItems.firstOrNull {
                it.lowercasedName == "bvid" && !it.value.isNullOrEmpty()
            }?.value
            if (bvid != null) return "/video/$bvid"

            val aid = parsed.queryItems.firstOrNull {
                (it.lowercasedName == "aid" || it.lowercasedName == "avid") &&
                    !it.value.isNullOrEmpty()
            }?.value
            if (aid != null) return "/video/av$aid"

            return null
        }

        private val preservedQueryNames = setOf(
            "p", "t", "start_progress", "start_progress_ms",
            "spm_id_from", "vd_source", "from_spmid", "from_source",
        )

        private fun filteredVideoQueryItems(items: List<BiliUrl.QueryItem>): List<BiliUrl.QueryItem> {
            val filtered = items.filter { it.lowercasedName in preservedQueryNames }
            return filtered.ifEmpty { null } ?: emptyList()
        }

        private fun canonicalLiveRoomID(parsed: BiliUrl): String? {
            val pathComponents = parsed.path.split("/").filter { it.isNotEmpty() }
            val numericFromPath = pathComponents.lastOrNull { it.all { c -> c.isDigit() } }
            if (numericFromPath != null) return numericFromPath

            val liveRoomKeys = setOf("room_id", "roomid", "id")
            return parsed.queryItems
                .firstOrNull { it.lowercasedName in liveRoomKeys }
                ?.value
                ?.trim()
                ?.takeIf { it.isNotEmpty() }
        }
    }
}

File: Sources/FocusCore/FocusNavigationPolicy.swift
import Foundation

public struct FocusNavigationPolicy: Sendable {
    public enum Decision: Equatable, Sendable {
        case allow
        case cancel
        case redirect(URL)
    }

    private static let blockedSchemes: Set<String> = [
        "bilibili",
        "bilibilihd",
        "bstar",
        "intent",
        "itms-apps",
        "itmss",
    ]
    private static let blockedHosts: Set<String> = [
        "app.bilibili.com",
    ]

    public init(settings _: FocusSettings = .defaults) {}

    public func decision(for url: URL, currentURL: URL? = nil) -> Decision {
        let canonicalURL = Self.canonicalWebURL(for: url)
        if canonicalURL != url {
            return .redirect(canonicalURL)
        }

        if shouldBlock(url: canonicalURL, currentURL: currentURL) {
            return .cancel
        }

        return .allow
    }

    public static func canonicalWebURL(for url: URL) -> URL {
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        guard let host = components?.host?.lowercased() else {
            return url
        }

        let path = url.path.lowercased()
        let isStandardBilibiliHost = host == "www.bilibili.com" || host == "bilibili.com" || host == "m.bilibili.com"

        if isStandardBilibiliHost && (path.hasPrefix("/video/") || path.hasPrefix("/bangumi/play/")) {
            components?.scheme = "https"
            components?.host = "www.bilibili.com"
            if path.hasPrefix("/video/"), components?.path.hasSuffix("/") == false {
                components?.path += "/"
            }
            if path.hasPrefix("/video/") {
                let filteredItems = filteredVideoQueryItems(from: components?.queryItems)
                components?.queryItems = filteredItems
            }
            return components?.url ?? url
        }

        if isStandardBilibiliHost && (path.hasPrefix("/blackboard/html5player.html") || path.hasPrefix("/blackboard/html5mobileplayer.html")) {
            if let videoPath = canonicalVideoPath(from: components) {
                let filteredItems = filteredVideoQueryItems(from: components?.queryItems)
                components?.scheme = "https"
                components?.host = "www.bilibili.com"
                components?.path = videoPath
                components?.queryItems = filteredItems
                components?.fragment = nil
                if components?.path.hasPrefix("/video/") == true, components?.path.hasSuffix("/") == false {
                    components?.path += "/"
                }
                return components?.url ?? url
            }
        }

        if isStandardBilibiliHost && path.hasPrefix("/opus/") {
            components?.host = "www.bilibili.com"
            return components?.url ?? url
        }

        if host == "t.bilibili.com" {
            let pathComponents = url.path.split(separator: "/")
            if pathComponents.count == 1, pathComponents[0].allSatisfy(\.isNumber) {
                components?.scheme = "https"
                components?.host = "www.bilibili.com"
                components?.path = "/opus/\(pathComponents[0])"
                return components?.url ?? url
            }
        }

        if host == "live.bilibili.com", let roomID = canonicalLiveRoomID(from: components) {
            components?.scheme = "https"
            components?.host = "live.bilibili.com"
            components?.path = "/\(roomID)"
            components?.queryItems = nil
            components?.fragment = nil
            return components?.url ?? url
        }

        return url
    }

    private func shouldBlock(url: URL, currentURL: URL?) -> Bool {
        let scheme = url.scheme?.lowercased() ?? ""
        if Self.blockedSchemes.contains(scheme) {
            return true
        }

        let host = url.host?.lowercased() ?? ""
        if Self.blockedHosts.contains(host) {
            return true
        }

        let lowercasedPath = url.path.lowercased()
        let lowercasedURL = url.absoluteString.lowercased()
        let isBilibiliFamilyHost = host.contains("bilibili.com") || host.contains("hdslb.com")
        let cameFromVideoPage = currentURL?.path.hasPrefix("/video/") == true
            || currentURL?.host?.contains("bilibili.com") == true && currentURL?.absoluteString.contains("/video/") == true

        if isBilibiliFamilyHost && lowercasedPath.contains("/download") {
            return true
        }

        if lowercasedURL.contains("openapp") || lowercasedURL.contains("launchapp") {
            return true
        }

        if cameFromVideoPage && isBilibiliFamilyHost && lowercasedURL.contains("download") {
            return true
        }

        return false
    }

    private static func canonicalVideoPath(from components: URLComponents?) -> String? {
        guard let queryItems = components?.queryItems else {
            return nil
        }

        if let bvid = queryItems.first(where: { $0.name.caseInsensitiveCompare("bvid") == .orderedSame })?.value,
           !bvid.isEmpty
        {
            return "/video/\(bvid)"
        }

        if let aid = queryItems.first(where: { $0.name.caseInsensitiveCompare("aid") == .orderedSame || $0.name.caseInsensitiveCompare("avid") == .orderedSame })?.value,
           !aid.isEmpty
        {
            return "/video/av\(aid)"
        }

        return nil
    }

    private static func filteredVideoQueryItems(from items: [URLQueryItem]?) -> [URLQueryItem]? {
        let preservedNames: Set<String> = [
            "p",
            "t",
            "start_progress",
            "start_progress_ms",
            "spm_id_from",
            "from_spmid",
            "from_source",
        ]

        let filtered = (items ?? []).filter { preservedNames.contains($0.name.lowercased()) }
        return filtered.isEmpty ? nil : filtered
    }

    private static func canonicalLiveRoomID(from components: URLComponents?) -> String? {
        let pathComponents = (components?.path ?? "")
            .split(separator: "/")
            .map(String.init)

        if let numericPathComponent = pathComponents.reversed().first(where: { !$0.isEmpty && $0.allSatisfy(\.isNumber) }) {
            return numericPathComponent
        }

        let queryItems = components?.queryItems ?? []
        let liveRoomKeys: Set<String> = [
            "room_id",
            "roomid",
            "id",
        ]

        return queryItems.first(where: { liveRoomKeys.contains($0.name.lowercased()) })?.value
            .flatMap { value in
                let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
                return trimmed.isEmpty ? nil : trimmed
            }
    }
}
